File: pori/observability/events.py
"""Normalized agent event stream.

A single event shape that every provider adapter emits and every renderer
(CLI, Cloud SSE, JSONL) consumes, so live UX and the replay/audit trail come
from the same source. See
``.agent/reference-studies/streaming-event-architecture.md``.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Dict

# --- event types ------------------------------------------------------------
RUN_START = "run_start"
RUN_END = "run_end"
STEP_START = "step_start"
STEP_END = "step_end"
TEXT_DELTA = "text_delta"  # visible answer prose, streamed
THINKING_DELTA = "thinking_delta"  # reasoning prose, streamed (may be dimmed)
TOOL_CALL_START = "tool_call_start"  # the instant the tool name is known
TOOL_CALL_END = "tool_call_end"  # after execution: success + result
LLM_RETRY = "llm_retry"  # API retrying/rate-limited (not "still thinking")
ACTIVITY_CHANGED = "activity_changed"  # model-authored intent for current work
PLAN_CHANGED = "plan_changed"  # host-observed normalized plan snapshot


@dataclass
class PoriEvent:
    """One normalized agent event.

    Providers emit ``type`` + ``payload``; the agent enriches with ``step`` when
    it forwards to consumers. ``payload`` is event-specific, e.g.
    ``{"text": "..."}`` for TEXT_DELTA or ``{"name": "answer"}`` for
    TOOL_CALL_START.
    """

    type: str
    payload: Dict[str, Any] = field(default_factory=dict)
    step: int = 0


class JsonlEventSink:
    """Append PoriEvents as JSON lines for replay/audit.

    A renderer-agnostic consumer: compose it alongside a live renderer so the
    same event stream that drives the UI also produces a diff-able trail. Opens
    the file per event (simple + crash-safe); runs are bounded so this is cheap.
    """

    def __init__(self, path: str):
        self.path = path

    def __call__(self, event: "PoriEvent") -> None:
        try:
            line = json.dumps(
                {"type": event.type, "payload": event.payload, "step": event.step}
            )
            with open(self.path, "a", encoding="utf-8") as fh:
                fh.write(line + "\n")
        except Exception:
            # Logging must never break a run.
            pass

File: apps/shared/src/events.ts
/**
 * The Aloy event contract — a 1:1 mirror of the Pori kernel's `PoriEvent`
 * (`pori/observability/events.py`). Every backend SSE frame decodes to one of
 * these; every surface renders from this single shape.
 */

export const RUN_START = "run_start";
export const RUN_END = "run_end";
export const STEP_START = "step_start";
export const STEP_END = "step_end";
export const TEXT_DELTA = "text_delta"; // visible answer prose, streamed
export const THINKING_DELTA = "thinking_delta"; // reasoning prose, streamed
export const TOOL_CALL_START = "tool_call_start"; // tool name known
export const TOOL_CALL_END = "tool_call_end"; // after execution: success + result
export const LLM_RETRY = "llm_retry"; // API retrying / rate-limited
export const CLARIFICATION_REQUEST = "clarification_request"; // ask_user w/ options

export type PoriEventType =
  | typeof RUN_START
  | typeof RUN_END
  | typeof STEP_START
  | typeof STEP_END
  | typeof TEXT_DELTA
  | typeof THINKING_DELTA
  | typeof TOOL_CALL_START
  | typeof TOOL_CALL_END
  | typeof LLM_RETRY
  | typeof CLARIFICATION_REQUEST;

/** One normalized agent event: `{ type, payload, step }`. */
export interface PoriEvent<P = Record<string, unknown>> {
  type: PoriEventType | string;
  payload: P;
  step: number;
}

// --- payload shapes (event-specific) ----------------------------------------
export interface TextDeltaPayload {
  text: string;
}

export interface ToolCallStartPayload {
  name: string;
  [key: string]: unknown;
}

export interface ToolCallEndPayload {
  name?: string;
  success?: boolean;
  result?: unknown;
  [key: string]: unknown;
}

/** A structured `ask_user` request the UI renders as buttons; answer it via
 *  `AloyClient.submitClarification(id, value)`. */
export interface ClarificationRequestPayload {
  type: typeof CLARIFICATION_REQUEST;
  id: string;
  question: string;
  options: string[];
}
